Ensure that control points are at least the capture distance away from each other during campaign generation

MetalStormGhost opened the pull request:
Added a method in frontline.py which ensures the frontline conflict is always located between the control points and cannot move behind either one, where we couldn't find route segments. This is done by making sure the frontline is always placed at least one segment away from either control point.

Resolves #1912

MetalStormGhost committed:
Added a method in frontline.py which ensures the frontline conflict is always located between the control points and cannot move behind either one, where we couldn't find route segments. This is done by making sure the frontline is always placed at least one segment away from either control point.

Resolves #1912

RndName commented:
Hey @DanAlbert can you take a look onto these changes? Not feeling confident enough to review them as i never ever touched some frontline things and this is the last fix one the list to 5.1.0 🚀

DanAlbert commented:
> where we couldn't find route segments

Why would this happen? It seems like either this is the bug, or maybe `attack_distance` is wrong?

> This is done by making sure the frontline is always placed at least one segment away from either control point.

There's no guarantee that there _are_ segments. A single segment link is valid.

MetalStormGhost committed:
Implemented a check in mizcampaignloader.py to check the distances between control points in the campaign definition. If the distance between the control points is shorter than the minimum distance (capture trigger distance of 3000 meters), an error is displayed.

Also, Frontline.point_from_a() will no longer raise an exception if a frontline Point was not found in any of the segment point_a:s. Instead, the Point is calculated from point_b, in reverse direction.

Reverted the addition of Frontline._keep_frontline_within_bounds() in revision 4323ce4dc31a261862b6e20430e6e0bc96200702.

MetalStormGhost commented:
Started to work on checking the minimum distance between control points during campaign generation. Discussion in Discord #dev https://discordapp.com/channels/595702951800995872/595706037449785364/929084322890268754

Now produces an error like this:
![image](https://user-images.githubusercontent.com/89945461/148640281-a3864a0a-ed27-437c-831c-29446a2b24cc.png)

It's still currently checking all control points against all others, instead of just evaluating the ones which are immediately connected with supply routes. I was thinking we might only want to check the connected points. Comments?

DanAlbert commented on game/campaignloader/mizcampaignloader.py:
```diff
@@ -283,6 +284,27 @@ def control_points(self) -> Dict[int, ControlPoint]:
+            for cp_index_a in control_points:
+                control_point_a = control_points[cp_index_a]
+                for cp_index_b in control_points:
+                    control_point_b = control_points[cp_index_b]
```
https://docs.python.org/3/library/itertools.html#itertools.product

```python
for a, b in itertools.product(control_points.values(), control_points.values()):
    ...
```

MetalStormGhost replied:
Modified as suggested.

DanAlbert commented on game/theater/frontline.py:
```diff
@@ -158,8 +158,11 @@ def point_from_a(self, distance: float) -> Point:
+        # point_b, with heading towards the point_a, distance of (attack_distance - remaining_dist) meters.
+        return segment.point_b.point_from_heading(
+            segment.attack_heading.degrees - 180,
+            segment.attack_distance - remaining_dist,
```
This is unrelated to the PR description and so belongs in a separate PR.

Beyond that

> Point was not found in any of the segment point_a:s

Why not? Isn't that what the check above is suppose to guarantee?

MetalStormGhost replied:
No, that condition can be encountered e.g when the opposing force frontline units have been completely wiped out. In that situation _position_distance() will return the total attack_distance for that specific route, which is beyond the last point_a of the route (being actually at the point_b of the last segment).

MetalStormGhost committed:
Cleaned up the control point distance check in MizCampaignLoader.

MetalStormGhost committed:
Revert "Also, Frontline.point_from_a() will no longer raise an exception if a frontline Point was not found in any of the segment point_a:s. Instead, the Point is calculated from point_b, in reverse direction."

This reverts frontline.py back to be69d17345b9a71186f6b057f94498bca17bbac6.

MetalStormGhost committed:
Reverted frontline.py back to c844c364fa8a544b8f542a63c90da7162539640f.

RndName commented:
PR got stale, shall this be closed?

MetalStormGhost commented:
> PR got stale, shall this be closed?

I think I had made both changes which Dan had requested and marked this as ready for review. Would you be willing to take a look at this?

DanAlbert commented:
We're closing stale PRs. If this is still needed, please update and reopen.
